client/server/bundler/external-modules.js

```diff
@@ -0,0 +1,56 @@
+/**
+ * External Dependencies
+ */
+const fs = require( 'fs' );
+const path = require( 'path' );
+const { builtinModules } = require( 'module' );
+
+function getModule( request ) {
+	const parts = request.split( '/' );
+	if ( parts[ 0 ].startsWith( '@' ) ) {
+		return parts[ 0 ] + '/' + parts[ 1 ];
+	}
+	return parts[ 0 ];
+}
+
+module.exports = class ExternalModulesWriter {
+	constructor( options ) {
+		this.options = {
+			path: './build',
+			filename: 'modules.json',
+			...options,
+		};
+	}
+
+	apply( compiler ) {
+		compiler.hooks.afterEmit.tapAsync( 'ExternalModulesWriter', ( compilation ) => {
+			const externalModules = new Set();
+
+			for ( const module of compilation.modules ) {
+				if ( ! module.external ) {
+					continue;
+				}
+
+				const requestModule = getModule( module.userRequest );
+
+				// native Node.js module, not in node_modules
+				if ( builtinModules.includes( requestModule ) ) {
+					continue;
+				}
+
+				// loading local file by relative path, not in node_modules
+				if ( requestModule.startsWith( './' ) || requestModule.startsWith( '../' ) ) {
+					continue;
+				}
+
+				externalModules.add( requestModule );
```

Comment:
Instead of generating a file with the module names that we resolve manually using our own implementation later, can we change this plugin to get the list of resolved modules by webpack? That way we guarantee they resolve to the same file.

If that's not possible (webpack is ignoring anything with `node_modules` in the path, so chances are they don't try to resolve them), we could at least use webpack's `enhanced-resolve` to resolve them here and don't pass that burden to the consumer of the list.

Reply:
webpack never really resolves these external modules. When it sees a request that matches the externals criteria, e.g., `import React from 'react'`, it doesn't interpret the `react` specifier at all, and copies it verbatim to the output: `require( 'react' )`. It's then up to Node.js to resolve that at runtime.

> we could at least use webpack's `enhanced-resolve` to resolve them here and don't pass that burden to the consumer of the list.

The knowledge about how would webpack resolve the request doesn't look very relevant: webpack never does that resolution, and the bundle contains simple `require( 'react' )` calls. It's up to the runtime to resolve them.

It therefore seems right that the list contains a simple `react` item, passing the burden of interpreting that to the consumer. The consumer needs to verify and ensure that the runtime will be able to resolve `react` into a real package. Webpack is no longer involved at all at that moment.

Also, if I tried to resolve `react` in the webpack plugin, what context directory should I use? The `react` package has been requested by many files inside the source tree, and webpack resolved them all to a single external. Should I resolve `react` against `client/my-sites/themes` or against `client/server/render`? Hard to tell, it's requested from both places.